contracts/upgradability/diamond/diamond.rs

```diff
@@ -0,0 +1,235 @@
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+pub use crate::{
+    ownable::*,
+    traits::diamond::*,
+};
+use brush::{
+    declare_storage_trait,
+    modifiers,
+    traits::{
+        Flush,
+        Hash,
+    },
+};
+use ink_env::call::{
+    DelegateCall,
+    ExecutionInput,
+    Selector as InkSelector,
+};
+use ink_prelude::vec::Vec;
+use ink_storage::Mapping;
+
+pub use derive::DiamondStorage;
+
+pub const STORAGE_KEY: [u8; 32] = ink_lang::blake2x256!("brush::DiamondData");
+
+// TODO: Add support of Erc165
+#[derive(Default, Debug)]
+#[brush::storage(STORAGE_KEY)]
+pub struct DiamondData {
+    pub ownable: OwnableData,
+    // selector mapped to its facet
+    pub selector_to_hash: Mapping<Selector, Hash>,
+    // facet mapped to all functions it supports
+    pub hash_to_selectors: Mapping<Hash, Vec<Selector>>,
+    // code hash of diamond contract for immutable functions
+    pub self_hash: Hash,
+}
+
+declare_storage_trait!(DiamondStorage, DiamondData);
+
+impl<T: DiamondStorage> OwnableStorage for T {
+    fn get(&self) -> &OwnableData {
+        &DiamondStorage::get(self).ownable
+    }
+
+    fn get_mut(&mut self) -> &mut OwnableData {
+        &mut DiamondStorage::get_mut(self).ownable
+    }
+}
+
+impl<T: DiamondStorage + Flush + DiamondCut> Diamond for T {
+    #[modifiers(only_owner)]
+    default fn diamond_cut(&mut self, diamond_cut: Vec<FacetCut>, init: Option<InitCall>) -> Result<(), DiamondError> {
+        self._diamond_cut(diamond_cut, init)
+    }
+}
+
+pub trait DiamondInternal {
+    fn _diamond_cut(&mut self, diamond_cut: Vec<FacetCut>, init: Option<InitCall>) -> Result<(), DiamondError>;
+
+    fn _fallback(&self) -> !;
+
+    fn _init_call(&self, call: InitCall) -> !;
+
+    fn _handle_replace_immutable(&mut self, hash: Hash) -> Result<(), DiamondError>;
+
+    fn _handle_replace_existing(&mut self, facet_cut: &FacetCut) -> Result<(), DiamondError>;
+
+    fn _handle_existing_selector(&mut self, selector: Selector);
+
+    fn _remove_facet(&mut self, code_hash: Hash);
+
+    fn _add_function(&mut self, code_hash: Hash, selector: Selector);
+
+    fn _emit_diamond_cut_event(&self, diamond_cut: &Vec<FacetCut>, init: &Option<InitCall>);
+}
+
+impl<T: DiamondStorage + Flush + DiamondCut> DiamondInternal for T {
+    default fn _diamond_cut(&mut self, diamond_cut: Vec<FacetCut>, init: Option<InitCall>) -> Result<(), DiamondError> {
```

Comment:
You need to re-write that function. The logic is more straightforward.

If the `selectors` is empty { you remove the facet }
else {
iterate over each selector:
if the selector already exists for the faucet.code_hash { skip }
else if the selector already exists for another code hash { return an error with another code hash. // The caller should remove that selector from another facet first. }
else { add a new selector. }

iterate over old facet(replaced) and remove all selectors that is not part fo a new facet
}

Reply:
we are missing the case when I want to add functions from existing facet and also remove some (I first added function 1,2 from facet A and then I want to add function 3 and remove function 1 from the same facet)

Reply:
> iterate over old facet(replaced) and remove all selectors that is not part fo a new facet

That part will do that)

Reply:
yes, I missed that lol